File: R_Files/Data_Visualizations_Census.R
---
title: "Strategic Implications"
author: "McAelan Remigio"
format: html
toc: true
embed-resources: TRUE
smooth-scroll: TRUE
---

# libraries
library(ggplot2)
library(dplyr)
library(tidyr)
library(stringr)
library(ggrepel)
library(purrr)
library(table1)
library(knitr)
library(rlang)


# load dataset
df <- read.csv("filepath", stringsAsFactors = FALSE)

# pie chart function
create_pie_chart <- function(df, column_name, title = "Pie Chart Responses") {
  
  # pull values and drop NA/blank
  vals <- df[[column_name]]
  vals <- vals[!is.na(vals) & vals != ""]
  
  if (length(vals) == 0) {
    message(paste0("No non-empty data in column '", column_name, "'."))
    return(invisible(NULL))
  }
  
  # robust counts
  df_counts <- dplyr::tibble(Choice = vals) |>
    dplyr::count(Choice, name = "Count") |>
    dplyr::arrange(dplyr::desc(Count)) |>
    dplyr::mutate(Percent = round(100 * Count / sum(Count), 1),
                  Choice = paste0(Choice, " - ", Count, " (", Percent, "%)"))
  
  ggplot(df_counts, aes(x = "", y = Count, fill = Choice)) +
    geom_bar(width = 1, stat = "identity", color = "black") +
    coord_polar("y", start = 0) +
    theme_void() +
    scale_fill_brewer(palette = "Set3") + 
    labs(title = title) +
    theme(
      plot.title = element_text(hjust = 0.5, size = 14, face = "bold"),
      legend.position = "right", 
      legend.title = element_blank(),
      legend.text = element_text(size = 10)
    )
}


# bar chart
create_bar_chart <- function(df, column_name, title = "Bar Chart Responses") {
  values <- na.omit(df[[column_name]])
  
  df_values <- data.frame(Value = values)
  
  p <- ggplot(df_values, aes(x = Value)) +
    geom_bar(fill = "steelblue") +
    labs(title = title, x = column_name, y = "Count") +
    theme_minimal() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1))
  
  return(p)
}
```

## Regular Pie Chart Function for Categorical
create_pie_chart(df, "vars_n", title="Title (optional)")

# If Skip Case Not Skipped and Change Variable Names
df = df %>% 
  dplyr::mutate(
    updated_var_1 = dplyr::case_when(
      stringr::str_detect(var_1, "1") ~ "1",
      stringr::str_detect(var_1, "2") ~ "2",
      stringr::str_detect(var_1, "3") ~ "3",
      stringr::str_detect(var_1, "4") ~ "4",
      stringr::str_detect(var_1, "5") ~ "5",
      stringr::str_detect(var_1, "6") ~ "6",
      stringr::str_detect(var_1, "Not Set") ~ "Not Set",
      TRUE ~ "Not Set"
    )
  )
create_pie_chart(df, "updated_var_1", title="New Title")

### String Column + Vector for Col Counts
df2 <- df %>%
  rename(feedback_col = old_feedback_col) %>%
  dplyr::select(feedback_col)

# named vector: names become column names, scanned for 
vector_list <- c(
  Name1         = "\\bname1\\b",
  Name2           = "\\bname2\\b",
  Name3         = "\\bname3\\b",
  ...              = "\\...\\b",
  NameN    = "\\bnamen\\b"
)

# NA
txt <- str_to_lower(coalesce(df2$feedback_col, ""))

vector_cols <- lapply(vector_list, function(p) {
  str_detect(txt, regex(p, ignore_case = TRUE))
})

vector_name <- df2 %>%
  bind_cols(as_tibble(vector_cols)) %>%
  mutate(any_vector_name = reduce(across(all_of(names(vector_list))), `|`))

# Tidy counts for each name (descending)
name_counts <- vector_name %>%
  summarise(across(all_of(names(vector_list)), ~ sum(.x, na.rm = TRUE))) %>%
  pivot_longer(everything(), names_to = "name", values_to = "count") %>%
  arrange(desc(count))

# Returns the counts from any data frame + text column
count_names <- function(data, text_col) {
  col_sym <- rlang::ensym(text_col)
  txt <- str_to_lower(coalesce(dplyr::pull(data, !!col_sym), ""))
  vector_cols <- lapply(vector_list, function(p) {
    str_detect(txt, regex(p, ignore_case = TRUE))
  })
  as_tibble(vector_cols) %>%
    summarise(across(everything(), ~ sum(.x, na.rm = TRUE))) %>%
    pivot_longer(everything(), names_to = "name", values_to = "count") %>%
    arrange(desc(count))
}

count_names(df2, feedback_col) %>%
  knitr::kable(align = c("l", "r"), caption = "Name Mentions Count")


table1::table1(~Name1:Name2:Name3:...:table1::setLabel(Name4, "Name 4"):any_vector_name, data = vector_name, caption = "Name Count from Feedback Column")
# name 4 example of setting label

# Filtering for characters
df_filter <- df %>%
  dplyr::filter(stringr::str_detect(tolower(as.character(var_n)), "string"))

df_filter_2 <- df_grad %>%
  dplyr::filter(!is.na(Experiential_Learning) & Experiential_Learning != "")

create_pie_chart(df_grad_el, "String", title = "title")

# Setup for Subsets
df_yes    <- df %>% filter(!is.na(vars) & vars != "")
df_no  <- df %>% filter(is.na(vars) | vars == "")

# flags
df_filter <- df_filter %>%
  mutate(
    insert_variable =
      (!is.na(name_of_one != "") |
      (!is.na(name_of_two != "") |
      (!is.na(name_of_three != "")
  ) %>%
  mutate(has_name_after_flag = ifelse(has_name_after, "Named", "None Named"))

df_el <- df %>% filter(!is.na(a_variable) & a_variable != "")

create_pie_chart(df_grad %>% filter(!is.na(variable) & variable != ""),
                 "has_name_after_flag",
                 title = "Title")
create_pie_chart(df_grad %>% filter(is.na(variable) | variable == ""),
                 "has_name_after_flag",
                 title = "Title (Non)")

File: Python_Files/Data_Visualizations_Census.py
"""
Title: Strategic Implications
Author: McAelan Remigio
"""
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import re
from plotly.express import pie as px_pie
from IPython.display import display

df = pd.read_csv("filepath.csv")

# Visualization Functions
def create_pie_chart(df, column_name, title="Pie Chart Responses"):
    vals = df[column_name].dropna()
    vals = vals[vals != ""]
    if vals.empty:
        print(f"No non-empty data in column '{column_name}'.")
        return None
      
    df_counts = (
        vals.value_counts()
        .reset_index()
        .rename(columns={"index": "Choice", column_name: "Count"})
    )
    df_counts["Percent"] = round(100 * df_counts["Count"] / df_counts["Count"].sum(), 1)
    df_counts["Label"] = df_counts.apply(
        lambda x: f"{x['Choice']} - {x['Count']} ({x['Percent']}%)", axis=1
    )

    plt.figure(figsize=(6, 6))
    plt.pie(df_counts["Count"], labels=df_counts["Label"], colors=sns.color_palette("Set3"))
    plt.title(title, fontsize=14, fontweight="bold")
    plt.show()

def create_bar_chart(df, column_name, title="Bar Chart Responses"):
    values = df[column_name].dropna()
    plt.figure(figsize=(8, 5))
    sns.countplot(x=values, color="steelblue")
    plt.title(title)
    plt.xticks(rotation=45, ha="right")
    plt.xlabel(column_name)
    plt.ylabel("Count")
    plt.tight_layout()
    plt.show()

# Example
# create_pie_chart(df, "vars_n", title="Title (optional)")

def update_var_column(df, var_name):
    def categorize(value):
        if pd.isna(value):
            return "Not Set"
        for i in range(1, 7):
            if str(i) in str(value):
                return str(i)
        return "Not Set"

    df[f"updated_{var_name}"] = df[var_name].apply(categorize)
    return df

# Example
# df = update_var_column(df, "var_1")
# create_pie_chart(df, "updated_var_1", title="New Title")

# Text Column + Vector Matching
def count_names(data, text_col, vector_list):
    txt = data[text_col].fillna("").str.lower()
    counts = {name: txt.str.contains(pattern, regex=True).sum() for name, pattern in vector_list.items()}
    name_counts = pd.DataFrame(list(counts.items()), columns=["name", "count"]).sort_values("count", ascending=False)
    display(name_counts)
    return name_counts

# Example vector patterns
vector_list = {
    "Name1": r"\bname1\b",
    "Name2": r"\bname2\b",
    "Name3": r"\bname3\b",
    # ...
    "NameN": r"\bnamen\b"
}

# Example
# df2 = df.rename(columns={"old_feedback_col": "feedback_col"})[["feedback_col"]]
# count_names(df2, "feedback_col", vector_list)

# Filters
def filter_contains(df, column, string):
    return df[df[column].astype(str).str.lower().str.contains(string.lower(), na=False)]

def filter_not_blank(df, column):
    return df[df[column].notna() & (df[column] != "")]

def filter_blank(df, column):
    return df[df[column].isna() | (df[column] == "")]
  
# Example
# df_filter = filter_contains(df, "var_n", "string")
# df_filter_2 = filter_not_blank(df, "Experiential_Learning")

# Subsets
def split_yes_no(df, column):
    df_yes = filter_not_blank(df, column)
    df_no = filter_blank(df, column)
    return df_yes, df_no
  
# Flags
def create_flag(df, name_cols):
    df = df.copy()
    df["has_name_after_flag"] = df[name_cols].notna().any(axis=1)
    df["has_name_after_flag"] = df["has_name_after_flag"].replace({True: "Named", False: "None Named"})
    return df

# Example 
# df_flagged = create_flag(df, ["name_of_one", "name_of_two", "name_of_three"])
# create_pie_chart(df_flagged, "has_name_after_flag", title="Title")
# df_el = filter_not_blank(df, "a_variable")
# create_pie_chart(df_el, "String", title="title")
# df_yes, df_no = split_yes_no(df, "vars")
# create_pie_chart(df_yes, "has_name_after_flag", title="Title")
# create_pie_chart(df_no, "has_name_after_flag", title="Title (Non)")
